### ISSUE TITLE
When Reader fails JobListeners are not being invoked

### ISSUE THREAD

**Issue by chazo** [EasyBatch/easybatch-framework#169]
This issue can be replicated by simply modifying the file name in [ParallelTutorialWithRecordDispatching](https://github.com/EasyBatch/easybatch-tutorials/blob/master/src/main/java/org/easybatch/tutorials/advanced/parallel/ParallelTutorialWithRecordDispatching.java) to one that does not exist. The test will not end gracefully, instead, it will hang waiting for the poison record which is set as job listener, and in this case it is never invoked

**Comment by benas** [comment]
Thank you for pointing out this issue.

Listeners life cycle is designed to start **after** initializing the reader. So when an exception occurs in this phase, the job listener is not invoked.

I've already fixed this issue, I'll push the fix asap and keep you informed.

**Comment by benas** [comment]
Hi,

I deployed version 4.1.0-SNAPSHOT containing the fix for this issue in maven central repo.

You can give it a try (see instruction [here](https://github.com/EasyBatch/easybatch-framework#current-version)) and keep me informed in case of any issue.

Kind regards
Mahmoud

**Comment by echang** [comment]
Hi Mahmoud,

Thanks for fixing this, I think I am running into something similar, could you point me the exact url to download this snapshot jar file under https://oss.sonatype.org/content/repositories/snapshots please? 

Thanks,
Erick

**Comment by benas** [comment]
the core jar is here:

https://oss.sonatype.org/content/repositories/snapshots/org/easybatch/easybatch-core/4.1.0-SNAPSHOT/

but you may want other modules too.. If you use maven, just add the following repository in your pom:

```
<repository>
    <id>ossrh</id>
    <url>https://oss.sonatype.org/content/repositories/snapshots</url>
</repository>
```

and declare the dependency to the snapshot version of easy batch:

```xml
<dependency>
   <groupId>org.easybatch</groupId>
   <artifactId>easybatch-core</artifactId>
   <version>4.1.0-SNAPSHOT</version>
</dependency>
```

Kind regards
Mahmoud

**Comment by echang** [comment]
Awesome, thanks Mahmoud!

Thanks,
Erick

**Comment by echang** [comment]
Hi Mahmoud,
I just gave it a try, looks like it is not the cure to my case.

I tried to build an alternative asynchronous pipeline by using jobs and blockingqueue as the buffer in between, so my job1 is database reading and dispatch records to the queue (queue 1) for the  job2 to read from and do some process there, then dispatch it to another queue (queue 2) for job3 to pick it up and do the work. I put the .jobListener((JobListener) new PoisonRecordBroadcaster(queue1)) when I build job1 and .jobListener((JobListener) new PoisonRecordBroadcaster(queue2)) in order for all the jobs to exit after all the records are processed but only job1 and job2 stops. I have multiple job2s and job3s running in parallel with a dedicated queue for each of them, and in job2 and job3 I read the queue in batch to process. I am wondering if I messed up anything here, so just want to check with you my idea of exiting the job is correct, or is there other mechanism to exit the jobs after all records are processed?

Thanks again,
Erick

**Comment by benas** [comment]
Hi Erick,

> I just gave it a try, looks like it is not the cure to my case.

The fix I made is to make sure that the job listener is invoked when an exception occurs during opening the record reader (see [test case](https://github.com/EasyBatch/easybatch-framework/blob/0e3ab9640f4f80b8ec97b56192468e2365b8dc2b/easybatch-core/src/test/java/org/easybatch/core/job/JobImplTest.java#L182)). This fixes the [parallel tutorial](https://github.com/EasyBatch/easybatch-tutorials/blob/master/src/main/java/org/easybatch/tutorials/advanced/parallel/ParallelTutorialWithRecordDispatching.java) when the input file does not exist as mentionned by @chazo . This should also fix the general use case when the listener should be invoked when not able to initialize the reader.

Your use case is quite different I think. May I ask you please to put all details of your use case in issue #166 ? It is a better place to discuss your scenario. It would be awesome to share your code in a gist or a github repo to best help you .

@chazo Do not hesitate to keep me informed if the fix is ok for you. Take your time, no worries :smile: 

Regards
Mahmoud

**Comment by chazo** [comment]
@benas works for me. For now I have a workaround 
```java
executorService.submit(() -> {
			final AtomicReference<JobReport> readJobReportRef = new AtomicReference<JobReport>(null);
			readJobReportRef.set(readerJob.call());
			// TODO remove this once Easybatch can call the job listener if a read fails			       
			poisonBroadcast.afterJobEnd(readJobReportRef.get());
		});
```
and I tested 4.1.0-SNAPSHOT without the above changes and it works fine. Do you have an estimated release date for 4.1.0?

**Comment by benas** [comment]
Good news! thank you for taking time to test.

I planned to release v4.1 by the end of march.. But if there are some others filed issues, I will release a bug fix version 4.0.1 quickly.

So don't hesitate to break the tool and file bugs!
